opentelemetry-sdk/src/attributes/set.rs

```diff
@@ -104,37 +105,47 @@ impl Eq for HashKeyValue {}
 ///
 /// This must implement [Hash], [PartialEq], and [Eq] so it may be used as
 /// HashMap keys and other de-duplication methods.
-#[derive(Clone, Default, Debug, Hash, PartialEq, Eq)]
-pub struct AttributeSet(BTreeSet<HashKeyValue>);
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct AttributeSet(Vec<HashKeyValue>, u64);
```

Comment:
Trying to see how caching the hash is improve speed.. I believe this is because it can now be re-used when we lookup timeseries in [here](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/metrics/internal/sum.rs#L34)?

It'd help if we were needing the hash calculation more than once. But we don't need it for more than once.. So in this PR, the hashcalculation is simply moved from [here](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/metrics/internal/sum.rs#L34) to the AttributeSet creation, so I am not able to see how that would have helped with the perf improvement! 

Or the perf gains are purely from the Btee->Vec switch?

Reply:
In other words, we should not make this change, as taking up more memory to store/cache the hash only benefits *if* there lookup/insert are needed more than once, and from my understanding, we don't have the lookup-more-than-once requirement..

Reply:
Putting a breakpoint on `Hash` shows there are several hidden hash calls within the hashmap.  For example:

![image](https://github.com/open-telemetry/opentelemetry-rust/assets/139393/86c2870a-714e-4efb-9592-d025a569d5ae)

So when the hash map grows, it has to re-hash all the time series in it.  So this maybe why we see a 500% increase in stress test performance but a bit muted (but still impactful) performance improvement in real life applications.  Precaching hashes improves this test case because we have 1k+ time series that sometimes need to be rehashed.  

Although eventually that should stabilize, and after a few thousand iterations I no longer see reserve calls requesting hashes, but I don't see a significant change in throughput.  

I did the following two runs:

```
mshapiro@matt-ms-hpz:~/code/opentelemetry-rust$ cargo run --release --package stress --bin metrics
   Compiling opentelemetry_sdk v0.21.1 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-sdk)
   Compiling opentelemetry-stdout v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-stdout)
   Compiling opentelemetry-appender-tracing v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-appender-tracing)
   Compiling stress v0.1.0 (/home/mshapiro/code/opentelemetry-rust/stress)
    Finished release [optimized] target(s) in 5.88s
     Running `target/release/metrics`
Number of threads: 6
Throughput: 5,525,800 iterations/sec
Throughput: 5,796,800 iterations/sec
Throughput: 5,553,600 iterations/sec
Throughput: 5,823,200 iterations/sec
Throughput: 5,942,600 iterations/sec
Throughput: 5,609,600 iterations/sec
^C
mshapiro@matt-ms-hpz:~/code/opentelemetry-rust$ cargo run --release --package stress --bin metrics
   Compiling opentelemetry_sdk v0.21.1 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-sdk)
   Compiling opentelemetry-appender-tracing v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-appender-tracing)
   Compiling opentelemetry-stdout v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-stdout)
   Compiling stress v0.1.0 (/home/mshapiro/code/opentelemetry-rust/stress)
    Finished release [optimized] target(s) in 5.86s
     Running `target/release/metrics`
Number of threads: 6
Throughput: 1,304,400 iterations/sec
Throughput: 1,314,800 iterations/sec
Throughput: 1,314,600 iterations/sec
Throughput: 1,158,400 iterations/sec
Throughput: 1,300,200 iterations/sec
Throughput: 1,332,600 iterations/sec
Throughput: 1,333,200 iterations/sec
```

The first run is with this PR as is, the 2nd run is commenting out the `impl Hash` block and adding `Hash` to the `derive()` attribute.

So pre-hashing is definitely providing an improvement on its own as far as  Ican tell.  This is backed up by profiling our actual application, as we see `Measure::call()` percentages go from 4.66% of our CPU time down to 2.30%.  I'm also pretty sure we saw a reduction in `hash_one` percentages but I didn't write them down in my notes.  

So I'm not totally sure where the extra hashing is coming from, but it does seem to add some benefits in real use cases.

Reply:
Good point. Looking into the code now I don't even see caching is going to help here. Even for the slight improvement, the implication of storing the cache could be considerable. We have a slightly different design in otel-cpp which helps as the lookup is not purely O(1) in the hot path.

Reply:
> So when the hash map grows, it has to re-hash all the time series in it. So this maybe why we see a 500% increase in stress test performance but a bit muted (but still impactful) performance improvement in real life applications. Precaching hashes improves this test case because we have 1k+ time series that sometimes need to be rehashed.

We need to prevent this nonetheless. We can ensure HashMap has enough size by initializing it correctly, so rehashed would be avoided. We know the max size (default 2000), so this should be relatively easy to achieve! Let us address this separately, as that is another nice improvement by avoiding reallocations!

My stress test show no change when I removed Hash custom implementation!

Reply:
Not sure what's going on then I guess.

For comparison I tested the Main branch stress test, then added precaching of the hashes (same mechanism) to the `BtreeSet` implementation and got 

``` 
mshapiro@matt-ms-hpz:~/code/opentelemetry-rust$ cargo run --release --package stress --bin metrics
   Compiling opentelemetry_sdk v0.21.1 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-sdk)
   Compiling opentelemetry-stdout v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-stdout)
   Compiling opentelemetry-appender-tracing v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-appender-tracing)
   Compiling stress v0.1.0 (/home/mshapiro/code/opentelemetry-rust/stress)
    Finished release [optimized] target(s) in 6.36s
     Running `target/release/metrics`
Number of threads: 6
Throughput: 1,109,400 iterations/sec
Throughput: 1,150,800 iterations/sec
Throughput: 1,153,000 iterations/sec
Throughput: 1,137,800 iterations/sec
Throughput: 1,135,000 iterations/sec
^C
mshapiro@matt-ms-hpz:~/code/opentelemetry-rust$ Throughput: 1,098,600 iterations/sec
cargo run --release --package stress --bin metrics
   Compiling opentelemetry_sdk v0.21.1 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-sdk)
   Compiling opentelemetry-stdout v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-stdout)
   Compiling opentelemetry-appender-tracing v0.2.0 (/home/mshapiro/code/opentelemetry-rust/opentelemetry-appender-tracing)
   Compiling stress v0.1.0 (/home/mshapiro/code/opentelemetry-rust/stress)
    Finished release [optimized] target(s) in 5.67s
     Running `target/release/metrics`
Number of threads: 6
Throughput: 3,475,400 iterations/sec
Throughput: 3,722,400 iterations/sec
Throughput: 4,138,400 iterations/sec
Throughput: 4,164,400 iterations/sec
Throughput: 4,053,600 iterations/sec
```

I've tried the same tests in windows (instead of just WSL) and got similar results (though a bit less of a gap).

Reply:
As another data point, I changed `ValueMap::new()` to do `HashMap::with_capacity()` to reduce resizing and rehashing.  That made the first iteration report to be slightly better than vanilla `main`, but not by much.  Every throughput report afterwards was within the same variance I get with main.

Reply:
rehashing impact cannot be tested using the current stress test, as there is no collect() triggered so the hashmap is never emptied, so after the initial fill-up, it wont need re-hash! 

Here's what I suggest:
1. We know from the tests that btree -> vec is definitely perf booster, so lets scope *this* PR to do just that.
2. For rehashing prevention, cache-the-hash, and avoid de-duplication logic, avoid sorting and more - lets do it separately on after other, so each PR is very focused and demonstrates clean perf gains.
(This PR already kick started a ton of ideas to improve perf!)

Thoughts?

Reply:
I'm not opposed to splitting it out.  I can do that tomorrow.

Reply:
Ok it's 12:30am, I went to go to sleep and it finally dawns on me why precalculating the hashing improves my throughput on multiple machines more than others.  So I had to get out my laptop to verify, and yes, I know what's going on now.

You are correct that we are only computing the hash once, but *where* we compute the hash is important!  Precomputing the hash the way I am in this PR (that I'm going to put in its own PR) incurs the cost at `AttributeSet::From()` time.  Not precomputing the hash incurs the cost at `Measure::call()` time.  

This is important because incurring the cost during the hash lookup means we are doing more work inside the mutex lock, and thus delaying all other `Measure::call()` operations.  Precomputing the hash allows each counter to compute the hash in parallel, and thus less work inside the hash.

This is why we were seeing such variance of if it was helping or not.  It all came down to thread counts for the throughput stress test.  I verified this by testing with and without precomputed hashes with 6 threads on my PC, then with and without with 2 threads (fyi, 1 thread is broken).  

On my personal laptop (and on battery):
* 6 threads precomputed: 3.5M
* 6 threads non-precomputed: 2.3M
* 2 threads prcomputed: 2.5M
* 2 threads non-precomputed: 2.5M

So precomputing the hash is a concurrency optimization.

Reply:
Awesome! It makes sense. Lets tackle it in the follow up PR.

If I learned anything over the last few years working in OTel - it is that there is no magic! Everything should be explainable. Thanks a lot!